### PR TITLE
Frustum culling

### PR DESCRIPTION
**Pull request by lokeshguddu**

### CODE REVIEW

**Comment by soumyajitdeb** [comment]
This update contains frustum culling only. However I have now managed to get both occlusion culling using occlusion queries (GLES 3.0) and frustum culling integrated together into a single update right now. 

Currently I'm doing a bit of cleanup and testing to make sure it runs properly. I shall be done by tomorrow eod. Please hold on to reviewing this one for the moment. Apologies for the confusion.

Thanks
Soumyajit

**Comment by JonShemitz** [comment]
Not formatted cleanly, and not 100% compatible with either monoscopic rendering or 2.0 memory management ... but also not an awful lot of code: we can probably adapt this.

File: GVRf/Framework/jni/engine/renderer/renderer.cpp [hunk 1]
```diff
@@ -59,17 +65,46 @@ void Renderer::renderCamera(std::shared_ptr<Scene> scene,
             std::shared_ptr<RenderData> render_data = (*it)->render_data();
             if (render_data != 0) {
                 if (render_data->material() != 0) {
-                    render_data_vector.push_back(render_data);
+                    // Check for frustum culling flag
+                    if(scene->get_frustum_culling())
+                    {
+                        // Frustum  Culling
+                        std::shared_ptr<Mesh> currentMesh = render_data->mesh();
+                        float* bounding_box_info = currentMesh->getBoundingBoxInfo();
```
Comment (thomasflynn) [inline GVRf/Framework/jni/engine/renderer/renderer.cpp]: it looks like getBoundingBoxInfo() iterates through all vertices of the Mesh each time it is called.  Also getBoundingBoxInfo() has a lot of redundant code with getBoundingBox().  Seems to me that when the Mesh is created, we could call getBoundingBox() to create the bounding box, store any extra info needed for culling, then just query that stored info when culling.  If there are any changes to the mesh, that would dirty the bounding box and force it to be recomputed.

File: GVRf/Framework/jni/gl/gl_texture.h [hunk 2]
```diff
@@ -32,7 +32,7 @@ class GLTexture {
         glBindTexture(target, id_);
         glTexParameteri(target, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
         glTexParameteri(target, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
-        glTexParameteri(target, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
+        glTexParameteri(target, GL_TEXTURE_MIN_FILTER, GL_LINEAR_MIPMAP_LINEAR);
```
Comment (thomasflynn) [inline GVRf/Framework/jni/gl/gl_texture.h]: I don't know if this was in the last check-in, if it was, i missed it.  linear_mipmap_linear is generally a lot more expensive than linear.  i don't feel comfortable defaulting to that.  but we should expose a mechanism for changing it.  Actually SIDIA also has a request for adjusting default texture parameters.

File: GVRf/Framework/jni/engine/renderer/renderer.cpp [hunk 3]
```diff
@@ -51,25 +51,108 @@ void Renderer::renderCamera(std::shared_ptr<Scene> scene,
...
+                            if(visible)
+                                render_data_vector.push_back(render_data);
+#if _GVRF_USE_GLES3_
+                            if (render_data->material() != 0)
+                            {
+                                if(scene->get_occlusion_culling())
+                                {
+                                    //Setup basic bounding box and material
+                                    std::shared_ptr<RenderData> bounding_box_render_data(new RenderData());
+                                    std::shared_ptr<Mesh> bounding_box_mesh = render_data->mesh()->getBoundingBox();
```
Comment (thomasflynn) [inline GVRf/Framework/jni/engine/renderer/renderer.cpp]: is this the only time getBoundingBox() is called?  it seems like calling it is necessary for the bounding box info in the frustum cull portion of the code.  i might be blind as to where else it is called though.

File: GVRf/Framework/jni/engine/renderer/renderer.cpp [hunk 4]
```diff
@@ -143,109 +228,162 @@ void Renderer::renderCamera(std::shared_ptr<Scene> scene,
             renderPostEffectData(texture_render_texture, post_effects.back(),
                     post_effect_shader_manager);
         }
-    } // flag checking
-}
 
-void Renderer::renderCamera(std::shared_ptr<Scene> scene,
```
Comment (thomasflynn) [inline GVRf/Framework/jni/engine/renderer/renderer.cpp]: did you really remove renderCamera?  or is diff just confused?  or are you not rebased to the monoscopic check-in?

**Comment by thomasflynn** [comment]
ok, yeah, it looks like you need to rebase.  it doesn't cherry-pick cleanly.  there's conflicts that need manual merging.  if you at least rebase this, we can consider merging this even with the issues i pointed out (i'll want to run it first though, which means i need to be able to cherry-pick it ;).

**Comment by chenchao1407** [comment]
Hi

Can you help do the manual rebase and then I can do the reviewing and test?

And also, please include the DCO in the commit comment.

Thanks in advance.
Sincerely
-Chao Chen

**Comment by soumyajitdeb** [comment]
This was closed as we had some problems with a manual rebase on our side. We've raised another pull request where things are merged correctly.

**Pull request closed without merging**